staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go

```diff
@@ -0,0 +1,1459 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"net/http/httptrace"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+func TestPerRequestWriteDeadlineWithWriteShouldReturnErrorAfterTimeout(t *testing.T) {
+	// This test documents the behavior of the per request write deadline
+	// using a standard net http server.
+	//
+	// scenario: we want to observe the circumstances under which the Write
+	// method of the ResponseWriter object returns an "i/o timeout" error:
+	//  - a) write deadline is set to 100ms
+	//  - b) the handler waits until the write deadline exceeds
+	//  - c) after timeout occurs, the handler keeps writing 1kB of data at a time
+	// we expect the following:
+	//  - d) client: expected to receive an error from the server
+	//  - e) server: the Write method of the ResponseWriter object should
+	//       return an "i/o timeout" error once its internal buffer is full
+	t.Parallel()
+
+	const deadline = 100 * time.Millisecond
+	tests := []struct {
+		protoMajor         int
+		waiter             waiter
+		clientErrVerifier  verifier
+		handlerErrVerifier verifier
+	}{
+		{
+			protoMajor: 1, // http/1x
+			// write timeout is set to 100ms, a wait of 5s should be
+			// enough to withstand flakes in CI.
+			waiter:             &waitWithDuration{after: 5 * time.Second},
+			clientErrVerifier:  wantContains{contains: "local error: tls: bad record MAC"},
```

Comment:
won't these break with the next Golang update? They look very low-level and might be changed at any time.

Reply:
Are you aware of a specific change coming? I just tested this with go version 1.23rc2, and got no failure.

Does go provide a more abstract way to test for this condition? Absent that, what would you recommend testing for here?